File: vm/memory.py
from dex.dexfile import CallSiteCache


class Memory:
    """
    This class implements an execution context for the DVM.
    It includes the execution state information like register values, field values, etc.

    Supports per-instance implementation. Separated out into a separate file
    for per-thread capabilities (will be implemented later)
    """

    def __init__(self, dex=None, fd=None):
        self.last_return = None
        self.last_exception = None

        self.method_instr_values = {}
        self.static_fields = {}
        self.instance_fields = {}

        self.call_site_cache = CallSiteCache()

        self.dex = dex
        self.fd = fd

File: rust/platypus-vm/src/memory.rs
/// VM execution context (memory/registers) — translates vm/memory.py

use std::collections::HashMap;

use super::value::Value;

/// Execution-context state, shared across instructions in one thread.
#[derive(Debug)]
pub struct Memory {
    /// Return value of the most recently executed instruction/call.
    pub last_return:   Option<Value>,
    /// Last exception raised during execution.
    pub last_exception: Option<Value>,

    /// Per-instruction result values (keyed by codepoint).
    pub method_instr_values: HashMap<u32, Value>,
    /// Static field values: field_idx → value.
    pub static_fields: HashMap<usize, Value>,
    /// Instance field values: field_idx → value.
    pub instance_fields: HashMap<usize, Value>,
}

impl Memory {
    pub fn new() -> Self {
        Memory {
            last_return:         None,
            last_exception:      None,
            method_instr_values: HashMap::new(),
            static_fields:       HashMap::new(),
            instance_fields:     HashMap::new(),
        }
    }
}

impl Default for Memory {
    fn default() -> Self {
        Self::new()
    }
}
